### PR TITLE
Feature/gh 179 bootstrap secure stack

### PR DESCRIPTION
**Pull request by laurentganne**
# Pull Request description

## Description of the change

Boostrapping a full stack setup, with secured communication between the Alient4Cloud yorc plugin, yorc server, and consul.

### What I did

Added the following options to yorc bootstrap :
  * `--insecure`: to install an insecure setup, default is secured
  * `--yorc_ca_passphrase` to provide a Certificate authority private key passphrase (required)
  * `--yorc_ca_key_file` to provide the path to a Certificate Authority private key, accessible locally. Optional, if not provided, a private key will be generated.
  * `--yorc_ca_pem_file` to provide the path to a PEM-encoded Certificate Authority, accessible locally. Optional, if not provided, a private key will be generated.
  * `--consul_encrypt_key` to provide a 16-bytes, Base64 encoded value of an encryption key used to encrypt Consul network traffic. Optional, if not provided, an encryption key will be generated.
  
Updated tosca types zip to include alien4Cloud and ystia forge components updates.
Updated single node and HA topology templates to deploy a secured setup.

### How to verify it

On an OpenStack setup for example, create a `bootstrap_inputs.yaml` file with this content using the plugin and yorc distributions needed for this feature,
where we rely on a pulp repository to get consul/terraform/jdk distributions to speed up the process,
replacing `pulp_repo_ip` by the pulp repository, `openstack_ip` by the openstack IP address, replacing openstack user/tenant/password/public network, and replacing the path to pem file and the ca_passphrase by a password you chose : 

```yaml
consul:
  download_url: http://pulp_repo_ip/ystia/yorc-deploy/consul_1.2.3_linux_amd64.zip
terraform:
  download_url: http://pulp_repo_ip/ystia/yorc-deploy/terraform_0.11.8_linux_amd64.zip
  plugins_download_urls:
  - http://pulp_repo_ip/ystia/yorc-deploy/terraform-provider-null_1.0.0_linux_amd64.zip
  - http://pulp_repo_ip/ystia/yorc-deploy/terraform-provider-consul_2.1.0_linux_amd64.zip
  - http://pulp_repo_ip/ystia/yorc-deploy/terraform-provider-openstack_1.9.0_linux_amd64.zip
  - http://pulp_repo_ip/ystia/yorc-deploy/terraform-provider-aws_1.36.0_linux_amd64.zip
  - http://pulp_repo_ip/ystia/yorc-deploy/terraform-provider-google_1.18.0_linux_amd64.zip
jdk:
  download_url: http://pulp_repo_ip/ystia/java-a4c/jdk-8u131-linux-x64.tar.gz
  version: 1.8.0-131-b11
yorc:
  download_url: https://ystia.jfrog.io/ystia/binaries/ystia/yorc/dist/feature/GH-179-bootstrap-secure-stack/yorc-3.2.0-SNAPSHOT.tgz
  private_key_file: /home/myuser/.ssh/yorc.pem
  ca_passphrase: changeme
yorc_plugin:
  download_url: https://ystia.jfrog.io/ystia/binaries/ystia/yorc-a4c-plugin/dist/feature/GH-82-SSL-config/alien4cloud-yorc-plugin-3.2.0-SNAPSHOT.zip
infrastructures:
  openstack:
    auth_url: http://openstack_ip:5000/v2.0
    default_security_groups:
    - openbar
    - default
    password: changeme
    private_network_name: private-test
    region: RegionOne
    tenant_name: mytenant
    user_name: myuser
address:
  floating_network_name: my-public-network
compute:
  image: "a460db41-e574-416f-9634-96f2862f10fe"
  flavor: 3
  key_pair: yorc
credentials:
  user: centos
```

Download, untar https://ystia.jfrog.io/ystia/binaries/ystia/yorc/dist/feature/GH-179-bootstrap-secure-stack/yorc-3.2.0-SNAPSHOT.tgz

Bootstrap a single node secure deployment running:
```bash
$ ./yorc bootstrap -v bootstrap_inputs.yaml -n test1
```

Once done, you can import into your Web browser the generated CA at `work/bootstrapResources/ca.pem` as a trusted Certficate Authority, or just login on Alien4Cloud and accept the security exception,
then check an orchestrator and location were configured.

Bootstrap a HA secure deployment running:
```bash
$ ./yorc bootstrap -v bootstrap_inputs.yaml -n test2 -d HA
```

Login on Alien4Cloud and check an orchestrator and location were configured. 

### Description for the changelog

Bootstrap a secure Yorc setup ([GH-179](https://github.com/ystia/yorc/issues/179))

## Applicable Issues

https://github.com/ystia/yorc/issues/179

### CODE REVIEW

**Comment by laurentganne** [comment]
> 
> 
> Looks Good
> Tested with GCP
> Maybe the passphrase could have a default value like "changeme" to make the conf simpler ?

Would prefer not to have a default value for security reasons, to make sure the user will provide a value and won't use something like changeme by inadvertance.

**Comment by stebenoist** [comment]
That's true.

File: commands/bootstrap/inputs.go [hunk 1]
```diff
@@ -631,6 +709,143 @@ func initializeInputs(inputFilePath, resourcesPath string, configuration config.
...
+		if err != nil {
+			return errors.Wrapf(err, "Failed to read CA key file %s", pConfig.CAKeyFile)
+		}
+		pConfig.CAKeyContent = string(data[:])
+	}
+
+	// Get CA or generate one
+	if pConfig.CAPEMContent == "" {
+
+		if pConfig.CAPEMFile == "" && askForInput {
```
Comment (loicalbertin) [inline commands/bootstrap/inputs.go]: Looks like we have duplicated code here

File: commands/bootstrap/setup.go [hunk 2]
```diff
@@ -75,7 +75,8 @@ func setupYorcServer(workingDirectoryPath string) error {
 		WorkersNumber:    inputValues.Yorc.WorkersNumber,
 		Infrastructures:  inputValues.Infrastructures,
 		Terraform: config.Terraform{
-			PluginsDir: workDirAbsolutePath,
+			PluginsDir:         workDirAbsolutePath,
+			KeepGeneratedFiles: true,
```
Comment (loicalbertin) [inline commands/bootstrap/setup.go]: Why?

File: commands/bootstrap/inputs.go (hunk 1 above)
Comment (laurentganne) [reply commands/bootstrap/inputs.go]: where ?

File: commands/bootstrap/setup.go (hunk 2 above)
Comment (laurentganne) [reply commands/bootstrap/setup.go]: To ease the bootstrap failure troubleshooting, did set ansible.KeepGeneratedRecipes too

**Pull request merged**